design/design-doc-resource-claim-definition.md

```diff
@@ -0,0 +1,158 @@
+* Owner: Nic Cope (@negz)
+* Reviewers: Crossplane Maintainers
+* Status: Draft
+
+## Background
+
+See the following issues this design doc intends to address:
+
+* https://github.com/crossplaneio/crossplane/issues/1105
+* https://github.com/crossplaneio/crossplane/issues/1106
+* https://github.com/crossplaneio/crossplane/issues/1107
+
+TODO(negz): Write this up.
+
+## Goals
+
+* Allow infrastructure operators to define claim kinds.
+* Allow a claim to be satisfied by multiple resources.
+* Surface claim to resource compatibility in Crossplane's API.
+
+TODO(negz): Write this up.
+
+## Proposal
+
+This document proposes the introduction of a new, cluster scoped, resource
+named `ResourceClaimDefinition`. A `ResourceClaimDefinition` builds on the
+existing `CustomResourceDefinition` concept, allowing an infrastructure
+operator to define arbitrary kinds of resource claim. Operators may define the
+schema for their claim, the resources it may bind to, and how it may dynamically
+provision those resources all without needing to write any code.
+
+```yaml
+apiVersion: core.crossplane.io/v1alpha1
+kind: ResourceClaimDefinition
+metadata:
+  name: mysqlinstance
+spec:
+  # A controller watching for ResourceClaimDefinition uses the group, names,
+  # version, and validation fields to create a CustomResourceDefinition.
+  group: database.crossplane.io
+  names:
+    kind: MySQLInstance
+    listKind: MySQLInstanceList
+    plural: mysqlinstances
+    singular: mysqlinstance
+  # We do not support multiple versions of the same ResourceClaimDefinition.
+  version: v1alpha1
+  # The validation field is used as an overlay when creating the CRD associated
+  # with this RCD. The base layer would contain all standard claim fields such
+  # that the operator did not need to respecify them all.
+  validation:
+    openAPIV3Schema:
+      description: A MySQLInstance is a portable resource claim that may be satisfied
+        by binding to a MySQL managed resource such as an AWS RDS instance or a GCP
+        CloudSQL instance.
+      properties:
+        spec:
+          properties:
+            engineVersion:
+              description: EngineVersion specifies the desired MySQL engine version,
+                e.g. 5.7.
+              enum:
+              - "5.6"
+              - "5.7"
+              type: string
+  # Satisfaction specifies one or more "resource groups" that can satisfy this
+  # claim kind. Typically each group would correspond to one or more resources
+  # from a single cloud provider, but you could also mix and match resources
+  # from different providers (e.g. a GKE cluster and an S3 bucket) if that was
+  # somehow useful to you.
+  satisfaction:
+  # GCP is a resource group that can satisfy this claim with a single resource.
+  - name: GCP
+    description: A MySQLInstance can be satisfied by a CloudSQLInstance.
+    resources:
+    - resource:
+        apiVersion: database.gcp.crossplane.io/v1beta1
+        kind: CloudSQLInstance
```

Comment:
Note that the CRD for these resources and classes are still published by their infra stacks as they are today, so this is a reference to an existing type, not a schema for one.

Also note there's no concept of "composite" managed resources in this design. I think the line between an infrastructure operator defined resource claim and an infrastructure operator defined managed resource becomes too blurry, so instead I prefer to allow a claim to reference more than one class and bind to more than one managed resource. We know how many and what specific kind of classes/resources are needed thanks to the RCD schema.